core/utils/async_utils.cpp

```diff
@@ -55,305 +54,102 @@ void busywait_mutex::unlock() noexcept {
-}
-
-void thread_pool::max_idle(size_t value) {
-  auto& state = *shared_state_;
-
-  {
-    std::lock_guard lock{state.lock};
-
-    max_idle_ = value;
-  }
-
-  state.cond.notify_all();  // wake any idle threads if they need termination
-}
-
-void thread_pool::max_idle_delta(int delta) {
-  auto& state = *shared_state_;
-
-  {
-    std::lock_guard lock{state.lock};
-    auto max_idle = max_idle_ + delta;
-
-    if (delta > 0 && max_idle < max_idle_) {
-      max_idle_ = std::numeric_limits<size_t>::max();
-    } else if (delta < 0 && max_idle > max_idle_) {
-      max_idle_ = std::numeric_limits<size_t>::min();
-    } else {
-      max_idle_ = max_idle;
-    }
-  }
-
-  state.cond.notify_all();  // wake any idle threads if they need termination
+template<bool UseDelay>
+ThreadPool<UseDelay>::ThreadPool(size_t threads,
+                                 std::basic_string_view<Char> name) {
+  start(threads, name);
 }
 
-size_t thread_pool::max_threads() const {
-  std::lock_guard lock{shared_state_->lock};
-
-  return max_threads_;
-}
-
-void thread_pool::max_threads(size_t value) {
-  auto& state = *shared_state_;
-
-  {
-    std::lock_guard lock{shared_state_->lock};
-
-    max_threads_ = value;
-
-    if (State::ABORT != state.state.load()) {
-      maybe_spawn_worker();
-    }
-  }
-
-  state.cond.notify_all();  // wake any idle threads if they need termination
-}
-
-void thread_pool::max_threads_delta(int delta) {
-  auto& state = *shared_state_;
-
-  {
-    std::lock_guard lock{state.lock};
-    auto max_threads = max_threads_ + delta;
-
-    if (delta > 0 && max_threads < max_threads_) {
-      max_threads_ = std::numeric_limits<size_t>::max();
-    } else if (delta < 0 && max_threads > max_threads_) {
-      max_threads_ = std::numeric_limits<size_t>::min();
-    } else {
-      max_threads_ = max_threads;
-    }
-
-    if (State::ABORT != state.state.load()) {
-      maybe_spawn_worker();
-    }
-  }
-
-  state.cond.notify_all();  // wake any idle threads if they need termination
-}
-
-bool thread_pool::run(std::function<void()>&& fn,
-                      clock_t::duration delay /*=0*/) {
-  if (!fn) {
-    return false;
-  }
-
-  auto& state = *shared_state_;
-
-  {
-    std::lock_guard lock{state.lock};
-
-    if (State::RUN != state.state.load()) {
-      return false;  // pool not active
-    }
-
-    queue_.emplace(std::move(fn), clock_t::now() + delay);
-
-    try {
-      maybe_spawn_worker();
-    } catch (...) {
-      if (0 == threads_.load()) {
-        // failed to spawn a thread to execute a task
-        queue_.pop();
-        throw;
+template<bool UseDelay>
+void ThreadPool<UseDelay>::start(size_t threads,
+                                 std::basic_string_view<Char> name) {
+  IRS_ASSERT(threads_.empty());
+  IRS_ASSERT(threads);
+  threads_.reserve(threads);
+  for (size_t i = 0; i != threads; ++i) {
+    threads_.emplace_back([&] {
+      if (!name.empty()) {
+        set_thread_name(name.data());
```

Comment:
It's not really valid, but it doesn't really matter for us, because we use only names that known in compile time.
Threads can be scheduled, but not runned, start function ended, name isn't valid, and only than threads run